modules/task/execution_test.go

```diff
@@ -0,0 +1,150 @@
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package task
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func init() {
+	_globalBackend = &InMemBackend{make(chan []byte, 2)}
+}
+
+func TestEnqueueNonFunction(t *testing.T) {
+	err := Enqueue("I am not a function")
+	assert.Error(t, err)
+	assert.Contains(t, err.Error(), "a func as input but was")
+}
+
+func TestEnqueueWithMultipleReturnValues(t *testing.T) {
+	fn := func() (string, error) { return "", nil }
+	err := Enqueue(fn)
+	assert.Error(t, err)
+	assert.Contains(t, err.Error(), "return only error but found")
+}
+
+func TestEnqueueFnDoesNotReturnError(t *testing.T) {
+	fn := func() string { return "" }
+	err := Enqueue(fn)
+	assert.Error(t, err)
+	assert.Contains(t, err.Error(), "return error but found")
+}
+
+func TestEnqueueFnWithMismatchedArgCount(t *testing.T) {
+	fn := func(s string) error { return nil }
+	err := Enqueue(fn)
+	assert.Error(t, err)
+	assert.Contains(t, err.Error(), "1 function arg(s) but found 0")
+}
+
+func TestEnqueueFnWithMismatchedArgType(t *testing.T) {
+	fn := func(s string) error { return nil }
+	err := Enqueue(fn, 1)
+	assert.Error(t, err)
+	assert.Contains(
+		t, err.Error(), "argument: 1 from type: int to type: string",
+	)
+}
+
+func TestEnqueueAndConsumeWithoutRegister(t *testing.T) {
+	fn := func(num float64) error { return nil }
+	err := Enqueue(fn, float64(1.0))
+	assert.NoError(t, err)
+	fnLookup.fnNameMap = make(map[string]interface{})
+	err = GlobalBackend().Consume()
+	assert.Error(t, err)
+	assert.Contains(
+		t, err.Error(), "\"go.uber.org/fx/modules/task.TestEnqueueAndConsumeWithoutRegister.func1\""+
+			" not found",
+	)
+}
+
+func TestEnqueueEncodingError(t *testing.T) {
+	fn := func(car Car) error { return nil }
+	fnLookup.fnNameMap[getFunctionName(fn)] = fn
+	err := Enqueue(fn, Car{})
+	assert.Error(t, err)
+	assert.Contains(t, err.Error(), "unable to encode the function")
+}
+
+func TestRunDecodeError(t *testing.T) {
+	err := Run([]byte{})
+	assert.Error(t, err)
+	assert.Contains(t, err.Error(), "unable to decode the message")
+}
+
+func TestEnqueueNoArgsFn(t *testing.T) {
+	err := Enqueue(NoArgs)
+	assert.NoError(t, err)
+	err = GlobalBackend().Consume()
+	assert.NoError(t, err)
+}
+
+func TestEnqueueSimpleFn(t *testing.T) {
+	err := Enqueue(SimpleWithError, "hello")
+	assert.NoError(t, err)
+	err = GlobalBackend().Consume()
+	assert.Error(t, err)
+	assert.Contains(t, err.Error(), "Simple error")
+}
+
+func TestEnqueueMapFn(t *testing.T) {
+	fn := func(map[string]string) error { return nil }
+	err := Enqueue(fn, make(map[string]string))
+	assert.NoError(t, err)
+	err = GlobalBackend().Consume()
+	assert.NoError(t, err)
+}
+
+func TestEnqueueComplexFnWithError(t *testing.T) {
```

Comment:
Good catch! Variadic arguments don't work. I think that's an ok restriction to have for now. Will add a task to follow up on that.